stan/math/prim/mat/fun/gp_exp_quad_cov.hpp

```diff
@@ -243,6 +246,186 @@ gp_exp_quad_cov(const std::vector<Eigen::Matrix<T_x1, Eigen::Dynamic, 1>> &x1,
+          typename Cond
+          = typename std::enable_if<is_constant<T_x>::value>::type>
+inline Eigen::MatrixXd gp_exp_quad_cov(const std::vector<T_x> &x,
+                                       const double sigma,
+                                       const double length_scale) {
+  check_positive("gp_exp_quad_cov", "magnitude", sigma);
+  check_positive("gp_exp_quad_cov", "length scale", length_scale);
+
+  size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  for (size_t n = 0; n < x.size(); ++n)
+    check_not_nan("gp_exp_quad_cov", "x", x[n]);
+
+  matrix_cl x_gpu(x);
+  matrix_cl cov_gpu = gp_exp_quad_cov(x_gpu, sigma, length_scale);
+  copy(cov, cov_gpu);  // NOLINT
+
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @param x std::vector of Eigen vectors of scalars.
+ * @param sigma marginal standard deviation or magnitude
+ * @param length_scale std::vector length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T = void>  // if this was non-templated overload or fully
+                              // specialized template, the compiler could
+                              // incorrectly resolve some called functions to
+                              // prim implementation instead of arr
+inline Eigen::MatrixXd gp_exp_quad_cov(
+    const std::vector<Eigen::VectorXd> &x, const double sigma,
+    const std::vector<double> &length_scale) {
+  check_positive_finite("gp_exp_quad_cov", "magnitude", sigma);
+  check_positive_finite("gp_exp_quad_cov", "length scale", length_scale);
+
+  size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  size_t l_size = length_scale.size();
+  check_size_match("gp_exp_quad_cov", "x dimension", x[0].size(),
+                   "number of length scales", l_size);
+
+  std::vector<Eigen::VectorXd> x_new = divide_columns(x, length_scale);
+
+  matrix_cl x_gpu(x_new);
+  matrix_cl cov_gpu = gp_exp_quad_cov(x_gpu, sigma, 1);
+  copy(cov, cov_gpu);  // NOLINT
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * This function is for the cross covariance matrix
+ * needed to compute posterior predictive density.
+ *
+ * @tparam T_x1 type of first std::vector of scalars
+ * @tparam T_x2 type of second std::vector of scalars
+ *    This function assumes each element of x1 and x2 are the same size.
+ * @tparam Cond condition
+ *
+ * @param x1 std::vector of elements that can be used in square distance
+ * @param x2 std::vector of elements that can be used in square distance
+ * @param sigma standard deviation
+ * @param length_scale length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T_x1, typename T_x2,
+          typename Cond = typename std::enable_if<
```

Comment:
I'm not sure if I understand exactley what are you suggesting, but I have simplified this a bit.

Reply:
The below is what I'm talking about, but it uses the C++17 fold expression stuff which idt we can use yet :(

I think there is a way to do that in C++11 but if we only use this in two or three places it may not be necessary rn

```
// checks whether types are all constant
template <typename... Args>
using is_all_constant = typename std::enable_if_t<is_constant_struct<Args>::value &&...>>
```

https://stackoverflow.com/a/48764324/2269255

Reply:
I think that could be something like this in c++11:
```
template<typename T, typename... Args>
struct is_constant{
    bool value = is_constant<T>::value && is_constant<Args ...>;
}
```

But I don't know whether we have enough use cases for this.

Reply:
Yeah if it's not used enough then let's not worry about it for now